[SILOpt] Hoist destroys for non-lexical alloc_stacks more aggressively.

nate-chandler opened the pull request:
Previously, when hoisting `destroy_addr`s for `alloc_stack`s, deinit barriers were always considered barriers to hoisting.  Here, that is changed so that hoisting only is barred when the `alloc_stack` whose `destroy_addr` is being hoisted is lexical.

Additionally, more non-lexical `alloc_stack`s are formed: only `alloc_box`es whose lifetimes are protected via a lexical borrow scope get promoted to lexical `alloc_stack`s.  This will become more significant when types and values are annotated `@_eagerMove`.

In order to do this safely, we must ensure that implicit lexical lifetimes are not removed.  Address function arguments with a convention other than Indirect_Inout have implicit lexical lifetimes: e.g. SSADestroyHoisting respects deinit barriers when hoisting the the destroy_addrs of such arguments.  Until we have `begin_borrow_addr`/`end_borrow_addr`, when inlining functions with such arguments, if the address passed to it has an `alloc_stack` as its storage base, mark that `alloc_stack` lexical.

nate-chandler committed:
[SSADestroyHoisting] Only lexical alloc_stacks respect deinit barriers.

nate-chandler committed:
[AllocBoxToStack] Lexical borrowees get lexical alloc_stacks.

Previously, whenever an alloc_box that was promoted to an alloc_stack,
the new alloc_stack would be lexical.  The result was that alloc_boxes
which didn't need (or explicitly didn't want, in the case of eager move
vars) received lexical alloc_stacks.

Here, only add the lexical flag to the new alloc_stack instruction if
any of the box's uses are `begin_borrow [lexical]`.  That way,
alloc_boxes end up having lexical alloc_stacks only if they were
"lexical alloc_boxes".

nate-chandler committed:
[SILInliner] Mark alloc_stacks lexical.

When inlining a function to which an alloc_stack is passed (via any
convention that's not all of mutating, exclusive, and inout), mark the
alloc_stack as lexical.  Otherwise, destroy_addrs could be hoisted
through the inlined function in a way inconsistent with how they were
hoisted through the callee prior to inlining.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please smoke test macOS platform

nate-chandler committed:
[TempRValueOpts] Opt some lexical alloc_stacks.

Previously, whenever an alloc_stack [lexical] was seen, optimization
bailed conservatively.

In the fullness of time, we should optimize an alloc_stack [lexical] so
long as:
(1) the root of the source of the store/copy_addr is lexical
(2) the range in which the alloc_stack [lexical] contains the value
    store'd/copy_addr'd into it (i.e. the range ending at
    destroy_addr/etc) is contained within the live range of that lexical
    root

Here, an incremental step in that direction is taken: if the base of the
accessed storage of the source of a copy_addr is a guaranteed function
argument, then we know
(1) the base is lexical--guaranteed function arguments are always
    lexical
(2) the range in which the alloc_stack contains the value copy_addr'd in
    is trivially contained within the range of that lexical root--the
    live range of a guaranteed function argument is the whole function

Added TODOs for the full optimization.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

atrick commented on lib/SILOptimizer/Transforms/AllocBoxToStack.cpp:
```diff
@@ -538,13 +540,31 @@ static bool rewriteAllocBoxAsAllocStack(AllocBoxInst *ABI) {
+    worklist.initializeRange(ABI->getUses());
+    while (auto *use = worklist.pop()) {
+      // See through mark_uninitialized instructions.
+      if (auto *mui = dyn_cast<MarkUninitializedInst>(use->getUser())) {
```
Hard-coding opcodes here (for mark_uninitialized) creates an implicit correctness dependency on SIL design. How will we find out if SILGen ever introduces new instructions in the same position as mark_uninitialized? Maybe we want a flag on alloc_box just at least as a verification mechanism?

nate-chandler replied:
Added verification that `begin_borrow [lexical]`s whose operand's type is a `SILBoxType` is either an `AllocBoxInst` or a `SILFunctionArgument` (looking through `copy_value`s, `begin_borrow`s, and `mark_uninitialized` instructions) in https://github.com/apple/swift/pull/60478/commits/fdb20e1b8dad188861cfe4ebeb038f4145952fe6 .

nate-chandler replied:
AllocBoxToStack also now looks through to the uses of non-lexical begin_borrows to properly handle the case of repeated inlining where the first callee's SILFunctionArgument is non-lexical but the second callee's SILFunctionArgument is lexical in https://github.com/apple/swift/pull/60478/commits/896a464ced0a7baaf30a202a58ed060ecfde4efa .

atrick reviewed (approved):
Looks great! Thanks.

nate-chandler committed:
[SILVerifier] Check lex borrows of box instances.

Require that lexical borrows of values which are instances of some
SILBoxType guard the lifetime of vars or closure captures; that's to
say, of AllocBoxInsts and SILFunctionArguments.

Look through a few instructions types to find those:
- copy_value - inserted by SILGen
- begin_borrow - inserted during inlining
- mark_uninitialized - inserted by SILGen

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility
